src/datafusion/scanner.rs

```diff
@@ -0,0 +1,195 @@
+    ) -> datafusion::error::Result<datafusion::physical_plan::SendableRecordBatchStream> {
+        let object_store = context
+            .runtime_env()
+            .object_store(&self.base_config.object_store_url)?;
+
+        let batch_size = context.session_config().batch_size();
+
+        let config = ZarrConfig::new(object_store)
+            .with_batch_size(batch_size)
+            .with_projection(self.base_config.projection.clone());
+
+        let opener = ZarrFileOpener::new(config);
+
+        let stream = FileStream::new(&self.base_config, partition, opener, &self.metrics)?;
+
+        Ok(Box::pin(stream) as SendableRecordBatchStream)
+    }
+
+    fn output_partitioning(&self) -> datafusion::physical_plan::Partitioning {
+        Partitioning::UnknownPartitioning(self.base_config.file_groups.len())
+    }
+
+    fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{error::Error, sync::Arc};
+
+    use datafusion::{
+        datasource::{listing::PartitionedFile, physical_plan::FileMeta},
+        execution::object_store::ObjectStoreUrl,
+    };
+    use futures::TryStreamExt;
+    use object_store::{local::LocalFileSystem, path::Path, ObjectMeta};
+
+    use crate::{
+        async_reader::{ZarrPath, ZarrReadAsync},
+        tests::get_test_v2_data_path,
+    };
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_open() -> Result<(), Box<dyn Error>> {
+        let local_fs = Arc::new(LocalFileSystem::new());
+
+        let test_data = get_test_v2_data_path("lat_lon_example.zarr".to_string());
+        let location = Path::from_filesystem_path(&test_data)?;
+
+        let file_meta = FileMeta {
+            object_meta: ObjectMeta {
+                location,
+                last_modified: chrono::Utc::now(),
+                size: 0,
+                e_tag: None,
+                version: None,
+            },
+            range: None,
+            extensions: None,
+        };
+
+        let zarr_path = ZarrPath::new(local_fs, file_meta.object_meta.location);
+        let schema = zarr_path.get_zarr_metadata().await?.arrow_schema()?;
+
+        let test_file = Path::from_filesystem_path(test_data)?;
+        let scan_config = FileScanConfig {
+            object_store_url: ObjectStoreUrl::local_filesystem(),
+            file_schema: Arc::new(schema.clone()),
+            file_groups: vec![vec![PartitionedFile::new(test_file.to_string(), 10)]],
+            statistics: Statistics::new_unknown(&schema),
+            projection: Some(vec![1, 2]),
+            limit: Some(10),
+            table_partition_cols: vec![],
+            output_ordering: vec![],
+        };
+
+        let scanner = ZarrScan::new(scan_config);
+
+        let session = datafusion::execution::context::SessionContext::new();
+
+        let batch_stream = scanner.execute(0, session.task_ctx())?;
```

Comment:
I haven't looked into data fusion for a while, and when I did there was a lot to go through so I remember not being 100% clear on how things worked, can you clarify what we get up to this point, by setting everything up and calling `execute`?

In this context, a "scan" simply reads all the data and produces batches? And I see that in this example, you're providing a limit of 10 rows to read, do you know how that's implemented? I took a look at the code you previously merged, I think we don't yet have a mechanism to push down a number of rows to read from a store (I guess that's kind of related to the batch size), is that right? So does the scan end up reading from the stream until it has a least 10 rows, then drops everything except the first 10 rows and returns that (as a record batch)?

Reply:
Sure, so a scan is what you describe it as... reading one or more files potentially with information that can be pushed down to the "physical" file reading and/or incorporated into the plan as a separate operator. Limit can be pushed down, or here it's redundantly incorporated into the scanning... e.g. https://github.com/apache/arrow-datafusion/blob/ad89ff82421e9c4670f4440dd5a6fa6fb55c40c3/datafusion/core/src/datasource/physical_plan/file_stream.rs#L405-L408  

Projection is similar in that in that it can operate as a node in the plan (and thus all fields could be wastefully read, and some just thrown out) or incorporated into the physical reading as is done in this PR.

Reply:
Sounds good, thanks for clarifying. Yeah that makes sense, with the current setup we can handle projections without wasteful reads, and we can incorporate something similar for limits later.

I'll review a bit more later tonight, partially because I want to understand how things work as early as possible in the process, but I don't want to hold up PRs for too long, I'll stamp it tonight and you can merge whenever you want after that.